File: go/indicators/johnehlers/discretefouriertransformspectrum/estimator.go
package discretefouriertransformspectrum

import "math"

// estimator implements the discrete Fourier transform power spectrum estimator.
// It is an unexported port of MBST's DiscreteFourierTransformSpectrumEstimator,
// used only by the DiscreteFourierTransformSpectrum indicator.
type estimator struct {
	length                          int
	spectrumResolution              int
	lengthSpectrum                  int
	maxOmegaLength                  int
	minPeriod                       float64
	maxPeriod                       float64
	isSpectralDilationCompensation  bool
	isAutomaticGainControl          bool
	automaticGainControlDecayFactor float64

	inputSeries          []float64
	inputSeriesMinusMean []float64
	spectrum             []float64
	period               []float64

	// Pre-computed trigonometric tables, size [lengthSpectrum][maxOmegaLength].
	// maxOmegaLength equals length (full-window DFT).
	frequencySinOmega [][]float64
	frequencyCosOmega [][]float64

	mean                float64
	spectrumMin         float64
	spectrumMax         float64
	previousSpectrumMax float64
}

// newEstimator creates a new discrete Fourier transform spectrum estimator.
func newEstimator(
	length int,
	minPeriod, maxPeriod float64,
	spectrumResolution int,
	isSpectralDilationCompensation bool,
	isAutomaticGainControl bool,
	automaticGainControlDecayFactor float64,
) *estimator {
	const twoPi = 2 * math.Pi

	lengthSpectrum := int((maxPeriod-minPeriod)*float64(spectrumResolution)) + 1
	maxOmegaLength := length

	e := &estimator{
		length:                          length,
		spectrumResolution:              spectrumResolution,
		lengthSpectrum:                  lengthSpectrum,
		maxOmegaLength:                  maxOmegaLength,
		minPeriod:                       minPeriod,
		maxPeriod:                       maxPeriod,
		isSpectralDilationCompensation:  isSpectralDilationCompensation,
		isAutomaticGainControl:          isAutomaticGainControl,
		automaticGainControlDecayFactor: automaticGainControlDecayFactor,
		inputSeries:                     make([]float64, length),
		inputSeriesMinusMean:            make([]float64, length),
		spectrum:                        make([]float64, lengthSpectrum),
		period:                          make([]float64, lengthSpectrum),
		frequencySinOmega:               make([][]float64, lengthSpectrum),
		frequencyCosOmega:               make([][]float64, lengthSpectrum),
	}

	result := float64(spectrumResolution)

	// Spectrum is evaluated from MaxPeriod down to MinPeriod with the configured resolution.
	for i := 0; i < lengthSpectrum; i++ {
		p := maxPeriod - float64(i)/result
		e.period[i] = p
		theta := twoPi / p

		sinRow := make([]float64, maxOmegaLength)
		cosRow := make([]float64, maxOmegaLength)

		for j := 0; j < maxOmegaLength; j++ {
			omega := float64(j) * theta
			sinRow[j] = math.Sin(omega)
			cosRow[j] = math.Cos(omega)
		}

		e.frequencySinOmega[i] = sinRow
		e.frequencyCosOmega[i] = cosRow
	}

	return e
}

// calculate fills mean, inputSeriesMinusMean, spectrum, spectrumMin, and spectrumMax
// from the current inputSeries contents.
func (e *estimator) calculate() {
	// Subtract the mean from the input series.
	mean := 0.0
	for i := 0; i < e.length; i++ {
		mean += e.inputSeries[i]
	}

	mean /= float64(e.length)

	for i := 0; i < e.length; i++ {
		e.inputSeriesMinusMean[i] = e.inputSeries[i] - mean
	}

	e.mean = mean

	// Evaluate the DFT power spectrum.
	e.spectrumMin = math.MaxFloat64
	if e.isAutomaticGainControl {
		e.spectrumMax = e.automaticGainControlDecayFactor * e.previousSpectrumMax
	} else {
		e.spectrumMax = -math.MaxFloat64
	}

	for i := 0; i < e.lengthSpectrum; i++ {
		sinRow := e.frequencySinOmega[i]
		cosRow := e.frequencyCosOmega[i]

		var sumSin, sumCos float64

		for j := 0; j < e.maxOmegaLength; j++ {
			sample := e.inputSeriesMinusMean[j]
			sumSin += sample * sinRow[j]
			sumCos += sample * cosRow[j]
		}

		s := sumSin*sumSin + sumCos*sumCos
		if e.isSpectralDilationCompensation {
			s /= e.period[i]
		}

		e.spectrum[i] = s

		if e.spectrumMax < s {
			e.spectrumMax = s
		}

		if e.spectrumMin > s {
			e.spectrumMin = s
		}
	}

	e.previousSpectrumMax = e.spectrumMax
}

File: ts/indicators/john-ehlers/discrete-fourier-transform-spectrum/estimator.ts
/** Internal discrete Fourier transform power spectrum estimator. Port of MBST's
 * DiscreteFourierTransformSpectrumEstimator used only by the DiscreteFourierTransformSpectrum
 * indicator; not exported from the module barrel. */
export class DiscreteFourierTransformSpectrumEstimator {
  public readonly length: number;
  public readonly spectrumResolution: number;
  public readonly lengthSpectrum: number;
  public readonly maxOmegaLength: number;
  public readonly minPeriod: number;
  public readonly maxPeriod: number;
  public readonly isSpectralDilationCompensation: boolean;
  public readonly isAutomaticGainControl: boolean;
  public readonly automaticGainControlDecayFactor: number;

  public readonly inputSeries: number[];
  public readonly inputSeriesMinusMean: number[];
  public readonly spectrum: number[];
  public readonly period: number[];

  // Pre-computed trigonometric tables, size [lengthSpectrum][maxOmegaLength].
  // maxOmegaLength equals length (full-window DFT).
  private readonly frequencySinOmega: number[][];
  private readonly frequencyCosOmega: number[][];

  public mean = 0;
  public spectrumMin = 0;
  public spectrumMax = 0;
  public previousSpectrumMax = 0;

  constructor(
    length: number,
    minPeriod: number,
    maxPeriod: number,
    spectrumResolution: number,
    isSpectralDilationCompensation: boolean,
    isAutomaticGainControl: boolean,
    automaticGainControlDecayFactor: number,
  ) {
    const twoPi = 2 * Math.PI;

    const lengthSpectrum = Math.trunc((maxPeriod - minPeriod) * spectrumResolution) + 1;
    const maxOmegaLength = length;

    this.length = length;
    this.spectrumResolution = spectrumResolution;
    this.lengthSpectrum = lengthSpectrum;
    this.maxOmegaLength = maxOmegaLength;
    this.minPeriod = minPeriod;
    this.maxPeriod = maxPeriod;
    this.isSpectralDilationCompensation = isSpectralDilationCompensation;
    this.isAutomaticGainControl = isAutomaticGainControl;
    this.automaticGainControlDecayFactor = automaticGainControlDecayFactor;

    this.inputSeries = new Array<number>(length).fill(0);
    this.inputSeriesMinusMean = new Array<number>(length).fill(0);
    this.spectrum = new Array<number>(lengthSpectrum).fill(0);
    this.period = new Array<number>(lengthSpectrum).fill(0);

    this.frequencySinOmega = new Array<number[]>(lengthSpectrum);
    this.frequencyCosOmega = new Array<number[]>(lengthSpectrum);

    const result = spectrumResolution;

    // Spectrum is evaluated from MaxPeriod down to MinPeriod with the configured resolution.
    for (let i = 0; i < lengthSpectrum; i++) {
      const p = maxPeriod - i / result;
      this.period[i] = p;
      const theta = twoPi / p;

      const sinRow = new Array<number>(maxOmegaLength);
      const cosRow = new Array<number>(maxOmegaLength);
      for (let j = 0; j < maxOmegaLength; j++) {
        const omega = j * theta;
        sinRow[j] = Math.sin(omega);
        cosRow[j] = Math.cos(omega);
      }
      this.frequencySinOmega[i] = sinRow;
      this.frequencyCosOmega[i] = cosRow;
    }
  }

  /** Fills mean, inputSeriesMinusMean, spectrum, spectrumMin, spectrumMax
   * from the current inputSeries contents. */
  public calculate(): void {
    // Subtract the mean from the input series.
    let mean = 0;
    for (let i = 0; i < this.length; i++) {
      mean += this.inputSeries[i];
    }
    mean /= this.length;

    for (let i = 0; i < this.length; i++) {
      this.inputSeriesMinusMean[i] = this.inputSeries[i] - mean;
    }
    this.mean = mean;

    // Evaluate the DFT power spectrum.
    this.spectrumMin = Number.MAX_VALUE;
    if (this.isAutomaticGainControl) {
      this.spectrumMax = this.automaticGainControlDecayFactor * this.previousSpectrumMax;
    } else {
      this.spectrumMax = -Number.MAX_VALUE;
    }

    for (let i = 0; i < this.lengthSpectrum; i++) {
      const sinRow = this.frequencySinOmega[i];
      const cosRow = this.frequencyCosOmega[i];

      let sumSin = 0;
      let sumCos = 0;

      for (let j = 0; j < this.maxOmegaLength; j++) {
        const sample = this.inputSeriesMinusMean[j];
        sumSin += sample * sinRow[j];
        sumCos += sample * cosRow[j];
      }

      let s = sumSin * sumSin + sumCos * sumCos;
      if (this.isSpectralDilationCompensation) {
        s /= this.period[i];
      }

      this.spectrum[i] = s;

      if (this.spectrumMax < s) this.spectrumMax = s;
      if (this.spectrumMin > s) this.spectrumMin = s;
    }

    this.previousSpectrumMax = this.spectrumMax;
  }
}
